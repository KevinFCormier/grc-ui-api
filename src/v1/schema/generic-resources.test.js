/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Generic Resources', () => {
  test('Should Report Error While Create Resources Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          createResources(resources:[{
            apiVersion: "mcm.ibm.com/v1alpha1",
            kind: "Application",
            metadata: {
              name: "testapp",
              labels: {
                deployable: "deployable01",
                hcmapp: "testapp",
              },
            },
            spec: {
              selector: {
                matchLabels: {
                  hcmapp: "testapp",
                },
              },
              componentKinds: [
                {
                  group: "mcm.ibm.com/v1alpha1",
                  kind: "PlacementPolicy",
                },
                {
                  group: "mcm.ibm.com/v1alpha1",
                  kind: "Deployable",
                },
              ],
            },
          }])
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Create Resources Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          createResources(resources:[{
            apiVersion: "mcm.ibm.com/v1alpha1",
            kind: "Application",
            metadata: {
              name: "testapp",
              namespace: "default",
              labels: {
                deployable: "deployable01",
                hcmapp: "testapp",
              },
            },
            spec: {
              selector: {
                matchLabels: {
                  hcmapp: "testapp",
                },
              },
              componentKinds: [
                {
                  group: "mcm.ibm.com/v1alpha1",
                  kind: "PlacementPolicy",
                },
                {
                  group: "mcm.ibm.com/v1alpha1",
                  kind: "Deployable",
                },
              ],
            },
          }])
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Update Cluster Labels', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          updateResourceLabels(resourceType:"HCMCompliance",namespace:"my-remote-cluster-1", name:"my-remote-cluster-1-xz", body:{cloud:"IBMs"}, selfLink:"apis/compliance.mcm.ibm.com/v1alpha1/namespaces/my-remote-cluster-1/compliances/compliance-xz", resourcePath:"/metadata/spec")
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves update Compliance Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          updateResource(resourceType: "HCMCompliance", namespace: "mcm", name: "test-compliance", selfLink: "/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/mcm/compliances/test-compliance", body:{
            apiVersion: "compliance.mcm.ibm.com/v1alpha1",
            kind: "Compliance",
            metadata: {
              name: "test-compliance",
              namespace: "mcm",
              description: "Instance descriptor for compliance resource",
            },
            spec: {
              clusterSelector: {
                matchNames: [
                  "mycluster",
                ],
              },
              runtimeRules: [
                {
                  apiVersion: "policy.mcm.ibm.com/v1alpha1",
                  kind: "Policy",
                  metadata: {
                    name: "test-policy-1",
                    description: "Instance descriptor for policy resource",
                  },
                  spec: {
                    remediationAction: "inform",
                    namespaces: {
                      include: [
                        "default",
                      ],
                      exclude: [
                        "kube*",
                      ],
                    },
                    roleTemplates: [
                      {
                        kind: "RoleTemplate",
                        apiVersion: "roletemplate.mcm.ibm.com/v1alpha1",
                        complianceType: "musthave",
                        metadata: {
                          namespace: "",
                          name: "role-xz-1",
                        },
                        selector: {
                          matchLabels: {
                            cloud: "IBM",
                          },
                        },
                        rules: [
                          {
                            complianceType: "musthave",
                            PolicyRule: {
                              apiGroups: [
                                "extensions",
                                "apps",
                              ],
                              resources: [
                                "deployments",
                              ],
                              verbs: [
                                "get",
                                "list",
                                "watch",
                                "create",
                                "delete",
                                "patch",
                              ],
                            },
                          },
                        ],
                      },
                    ],
                  },
                },
              ],
            },
          })
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});

