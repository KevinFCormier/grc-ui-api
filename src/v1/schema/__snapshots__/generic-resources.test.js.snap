// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic Resources Correctly Resolves Create Resources Mutation 1`] = `
Object {
  "data": Object {
    "createResources": Object {
      "errors": Array [],
      "result": Array [
        Object {
          "apiVersion": "mcm.ibm.com/v1alpha1",
          "kind": "Application",
          "metadata": Object {
            "annotations": Object {
              "mcm.ibm.com/user-group": "c3lzdGVtOmF1dGhlbnRpY2F0ZWQ=",
              "mcm.ibm.com/user-identity": "aHR0cHM6Ly9teWNsdXN0ZXIuaWNwOjk0NDMvb2lkYy9lbmRwb2ludC9PUCNhZG1pbg==",
            },
            "creationTimestamp": "2018-09-07T14:19:02Z",
            "labels": Object {
              "deployable": "deployable01",
              "hcmapp": "testapp",
            },
            "name": "testapp",
            "namespace": "default",
            "resourceVersion": "59905",
            "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/default/applications/testapp",
            "uid": "f849165a-b2a8-11e8-bd43-b69970856045",
          },
          "spec": Object {
            "componentKinds": Array [
              Object {
                "group": "mcm.ibm.com/v1alpha1",
                "kind": "PlacementPolicy",
              },
              Object {
                "group": "mcm.ibm.com/v1alpha1",
                "kind": "Deployable",
              },
            ],
            "descriptor": Object {},
            "selector": Object {
              "matchLabels": Object {
                "hcmapp": "testapp",
              },
            },
          },
          "status": Object {
            "Deployable": Object {
              "metadata": Object {
                "creationTimestamp": null,
              },
              "spec": Object {
                "deployer": Object {
                  "helm": Object {},
                },
              },
              "status": Object {},
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Generic Resources Correctly Resolves Update Cluster Labels 1`] = `
Object {
  "data": Object {
    "updateResourceLabels": Object {
      "apiVersion": "compliance.mcm.ibm.com/v1alpha1",
      "kind": "Compliance",
      "metadata": Object {
        "annotations": Object {
          "seed-generation": "1",
        },
        "creationTimestamp": "2019-01-02T19:42:12Z",
        "finalizers": Array [
          "finalizer.mcm.ibm.com",
        ],
        "generation": 1,
        "name": "compliance-all",
        "namespace": "mcm",
        "resourceVersion": "5533372",
        "selfLink": "/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/mcm/compliances/compliance-all",
        "uid": "7fc1f4a2-0ec6-11e9-8fd0-0ebe277f4f9c",
      },
      "spec": Object {
        "runtime-rules": Array [
          Object {
            "apiVersion": "policy.mcm.ibm.com/v1alpha1",
            "kind": "Policy",
            "metadata": Object {
              "creationTimestamp": null,
              "labels": Object {
                "hipaa": "true",
              },
              "name": "policy-all",
            },
            "spec": Object {
              "complianceType": "musthave",
              "namespaces": Object {
                "exclude": Array [
                  "kube*",
                ],
                "include": Array [
                  "default",
                ],
              },
              "object-templates": Array [
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "Pod",
                    "metadata": Object {
                      "name": "nginx",
                    },
                    "spec": Object {
                      "containers": Array [
                        Object {
                          "image": "nginx:1.7.9",
                          "name": "nginx",
                          "ports": Array [
                            Object {
                              "containerPort": 80,
                            },
                          ],
                        },
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "Namespace",
                    "metadata": Object {
                      "labels": Object {
                        "name": "production",
                      },
                      "name": "production",
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "rbac.authorization.k8s.io/v1",
                    "kind": "RoleBinding",
                    "metadata": Object {
                      "name": "operate-pods-rolebinding",
                      "namespace": "default",
                    },
                    "roleRef": Object {
                      "apiGroup": "rbac.authorization.k8s.io",
                      "kind": "Role",
                      "name": "operator",
                    },
                    "subjects": Array [
                      Object {
                        "apiGroup": "rbac.authorization.k8s.io",
                        "kind": "User",
                        "name": "jane",
                      },
                    ],
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "policy/v1beta1",
                    "kind": "PodSecurityPolicy",
                    "metadata": Object {
                      "annotations": Object {
                        "seccomp.security.alpha.kubernetes.io/allowedProfileNames": "*",
                      },
                      "name": "privileged-mcm",
                    },
                    "spec": Object {
                      "allowPrivilegeEscalation": true,
                      "allowedCapabilities": Array [
                        "*",
                      ],
                      "fsGroup": Object {
                        "rule": "RunAsAny",
                      },
                      "hostIPC": true,
                      "hostNetwork": true,
                      "hostPID": true,
                      "hostPorts": Array [
                        Object {
                          "max": 65535,
                          "min": 0,
                        },
                      ],
                      "privileged": true,
                      "runAsUser": Object {
                        "rule": "RunAsAny",
                      },
                      "seLinux": Object {
                        "rule": "RunAsAny",
                      },
                      "supplementalGroups": Object {
                        "rule": "RunAsAny",
                      },
                      "volumes": Array [
                        "*",
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "networking.k8s.io/v1",
                    "kind": "NetworkPolicy",
                    "metadata": Object {
                      "name": "allow-all-mcm",
                    },
                    "spec": Object {
                      "ingress": Array [
                        Object {},
                      ],
                      "podSelector": Object {},
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
                Object {
                  "complianceType": "musthave",
                  "objectDefinition": Object {
                    "apiVersion": "v1",
                    "kind": "LimitRange",
                    "metadata": Object {
                      "name": "mem-limit-range",
                    },
                    "spec": Object {
                      "limits": Array [
                        Object {
                          "default": Object {
                            "memory": "512Mi",
                          },
                          "defaultRequest": Object {
                            "memory": "256Mi",
                          },
                          "type": "Container",
                        },
                      ],
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
              "remediationAction": "enforce",
              "role-templates": Array [
                Object {
                  "apiVersion": "roletemplate.mcm.ibm.com/v1alpha1",
                  "complianceType": "musthave",
                  "metadata": Object {
                    "creationTimestamp": null,
                    "name": "operator-role",
                  },
                  "rules": Array [
                    Object {
                      "complianceType": "musthave",
                      "policyRule": Object {
                        "apiGroups": Array [
                          "extensions",
                          "apps",
                        ],
                        "resources": Array [
                          "deployments",
                        ],
                        "verbs": Array [
                          "get",
                          "list",
                          "watch",
                          "create",
                          "delete",
                          "patch",
                        ],
                      },
                    },
                  ],
                  "selector": Object {
                    "matchLabels": Object {
                      "hipaa": "true",
                    },
                  },
                  "status": Object {
                    "Validity": Object {},
                  },
                },
              ],
            },
            "status": Object {},
          },
        ],
      },
      "status": Object {
        "placementBindings": Array [
          "binding-xz",
        ],
        "placementPolicies": Array [
          "placement-xz",
        ],
        "status": Object {
          "cluster1": Object {
            "aggregatePoliciesStatus": Object {
              "policy-all": Object {
                "compliant": "Compliant",
                "valid": true,
              },
            },
            "clustername": "cluster1",
            "compliant": "Compliant",
          },
          "cluster2": Object {
            "aggregatePoliciesStatus": Object {
              "policy-all": Object {
                "compliant": "Compliant",
                "valid": true,
              },
            },
            "clustername": "cluster2",
            "compliant": "Compliant",
          },
          "cluster3": Object {
            "aggregatePoliciesStatus": Object {
              "policy-all": Object {
                "compliant": "Compliant",
                "valid": true,
              },
            },
            "clustername": "cluster3",
            "compliant": "Compliant",
          },
        },
      },
    },
  },
}
`;

exports[`Generic Resources Correctly Resolves update Compliance Mutation 1`] = `
Object {
  "data": Object {
    "updateResource": Object {
      "apiVersion": "compliance.mcm.ibm.com/v1alpha1",
      "items": Array [
        Object {
          "apiVersion": "compliance.mcm.ibm.com/v1alpha1",
          "kind": "Compliance",
          "metadata": Object {
            "annotations": Object {
              "seed-generation": "1",
            },
            "creationTimestamp": "2019-01-02T19:42:12Z",
            "finalizers": Array [
              "finalizer.mcm.ibm.com",
            ],
            "generation": 1,
            "name": "compliance-xz",
            "namespace": "mcm",
            "resourceVersion": "5533372",
            "selfLink": "/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/mcm/compliances/compliance-xz",
            "uid": "7fc1f4a2-0ec6-11e9-8fd0-0ebe277f4f9c",
          },
          "spec": Object {
            "runtime-rules": Array [
              Object {
                "apiVersion": "policy.mcm.ibm.com/v1alpha1",
                "kind": "Policy",
                "metadata": Object {
                  "creationTimestamp": null,
                  "labels": Object {
                    "hipaa": "true",
                  },
                  "name": "policy-all",
                },
                "spec": Object {
                  "complianceType": "musthave",
                  "namespaces": Object {
                    "exclude": Array [
                      "kube*",
                    ],
                    "include": Array [
                      "default",
                    ],
                  },
                  "object-templates": Array [
                    Object {
                      "complianceType": "musthave",
                      "objectDefinition": Object {
                        "apiVersion": "v1",
                        "kind": "Pod",
                        "metadata": Object {
                          "name": "nginx",
                        },
                        "spec": Object {
                          "containers": Array [
                            Object {
                              "image": "nginx:1.7.9",
                              "name": "nginx",
                              "ports": Array [
                                Object {
                                  "containerPort": 80,
                                },
                              ],
                            },
                          ],
                        },
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                    Object {
                      "complianceType": "musthave",
                      "objectDefinition": Object {
                        "apiVersion": "v1",
                        "kind": "Namespace",
                        "metadata": Object {
                          "labels": Object {
                            "name": "production",
                          },
                          "name": "production",
                        },
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                    Object {
                      "complianceType": "musthave",
                      "objectDefinition": Object {
                        "apiVersion": "rbac.authorization.k8s.io/v1",
                        "kind": "RoleBinding",
                        "metadata": Object {
                          "name": "operate-pods-rolebinding",
                          "namespace": "default",
                        },
                        "roleRef": Object {
                          "apiGroup": "rbac.authorization.k8s.io",
                          "kind": "Role",
                          "name": "operator",
                        },
                        "subjects": Array [
                          Object {
                            "apiGroup": "rbac.authorization.k8s.io",
                            "kind": "User",
                            "name": "jane",
                          },
                        ],
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                    Object {
                      "complianceType": "musthave",
                      "objectDefinition": Object {
                        "apiVersion": "policy/v1beta1",
                        "kind": "PodSecurityPolicy",
                        "metadata": Object {
                          "annotations": Object {
                            "seccomp.security.alpha.kubernetes.io/allowedProfileNames": "*",
                          },
                          "name": "privileged-mcm",
                        },
                        "spec": Object {
                          "allowPrivilegeEscalation": true,
                          "allowedCapabilities": Array [
                            "*",
                          ],
                          "fsGroup": Object {
                            "rule": "RunAsAny",
                          },
                          "hostIPC": true,
                          "hostNetwork": true,
                          "hostPID": true,
                          "hostPorts": Array [
                            Object {
                              "max": 65535,
                              "min": 0,
                            },
                          ],
                          "privileged": true,
                          "runAsUser": Object {
                            "rule": "RunAsAny",
                          },
                          "seLinux": Object {
                            "rule": "RunAsAny",
                          },
                          "supplementalGroups": Object {
                            "rule": "RunAsAny",
                          },
                          "volumes": Array [
                            "*",
                          ],
                        },
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                    Object {
                      "complianceType": "musthave",
                      "objectDefinition": Object {
                        "apiVersion": "networking.k8s.io/v1",
                        "kind": "NetworkPolicy",
                        "metadata": Object {
                          "name": "allow-all-mcm",
                        },
                        "spec": Object {
                          "ingress": Array [
                            Object {},
                          ],
                          "podSelector": Object {},
                        },
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                    Object {
                      "complianceType": "musthave",
                      "objectDefinition": Object {
                        "apiVersion": "v1",
                        "kind": "LimitRange",
                        "metadata": Object {
                          "name": "mem-limit-range",
                        },
                        "spec": Object {
                          "limits": Array [
                            Object {
                              "default": Object {
                                "memory": "512Mi",
                              },
                              "defaultRequest": Object {
                                "memory": "256Mi",
                              },
                              "type": "Container",
                            },
                          ],
                        },
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                  ],
                  "remediationAction": "enforce",
                  "role-templates": Array [
                    Object {
                      "apiVersion": "roletemplate.mcm.ibm.com/v1alpha1",
                      "complianceType": "musthave",
                      "metadata": Object {
                        "creationTimestamp": null,
                        "name": "operator-role",
                      },
                      "rules": Array [
                        Object {
                          "complianceType": "musthave",
                          "policyRule": Object {
                            "apiGroups": Array [
                              "extensions",
                              "apps",
                            ],
                            "resources": Array [
                              "deployments",
                            ],
                            "verbs": Array [
                              "get",
                              "list",
                              "watch",
                              "create",
                              "delete",
                              "patch",
                            ],
                          },
                        },
                      ],
                      "selector": Object {
                        "matchLabels": Object {
                          "hipaa": "true",
                        },
                      },
                      "status": Object {
                        "Validity": Object {},
                      },
                    },
                  ],
                },
                "status": Object {},
              },
            ],
          },
          "status": Object {
            "placementBindings": Array [
              "binding-xz",
            ],
            "placementPolicies": Array [
              "placement-xz",
            ],
            "status": Object {
              "cluster1": Object {
                "aggregatePoliciesStatus": Object {
                  "policy-all": Object {
                    "compliant": "Compliant",
                    "valid": true,
                  },
                },
                "clustername": "cluster1",
                "compliant": "Compliant",
              },
              "cluster2": Object {
                "aggregatePoliciesStatus": Object {
                  "policy-all": Object {
                    "compliant": "Compliant",
                    "valid": true,
                  },
                },
                "clustername": "cluster2",
                "compliant": "Compliant",
              },
              "cluster3": Object {
                "aggregatePoliciesStatus": Object {
                  "policy-all": Object {
                    "compliant": "Compliant",
                    "valid": true,
                  },
                },
                "clustername": "cluster3",
                "compliant": "Compliant",
              },
            },
          },
        },
      ],
      "kind": "ComplianceList",
      "metadata": Object {
        "continue": "",
        "resourceVersion": "4401288",
        "selfLink": "/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/mcm/compliances",
      },
    },
  },
}
`;

exports[`Generic Resources Should Report Error While Create Resources Mutation 1`] = `
Object {
  "data": Object {
    "createResources": Object {
      "errors": Array [
        Object {
          "message": "Namespace not found in the template",
        },
      ],
    },
  },
}
`;
